File: go/jzoffer/translateNum_46.go
package jzoffer

import "strconv"

// translateNum 翻译 num 成字符串 有几种方法
// 0 - a 1 - b ………… 25 - z
func translateNum(num int) int {
	str := strconv.Itoa(num)
	// dp[i] 表示 可以翻译成几种
	dp := make([]int, len(str)+1)
	dp[len(str)] = 1
	for i := len(str) - 1; i >= 0; i-- {
		for j := i + 1; j <= min(i+2, len(str)); j++ {
			if canTranslate(str[i:j]) {
				dp[i] += dp[j]
			}
		}
	}
	return dp[0]
}

func canTranslate(str string) bool {
	if len(str) > 1 && str[0] == '0' {
		return false
	}
	i, _ := strconv.Atoi(str)
	return i >= 0 && i <= 25
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: java/jzoffer/TranslateNum_46.java
public class TranslateNum_46 {

    public int translateNum(int num) {
        String str = String.valueOf(num);
        int[] dp = new int[str.length() + 1];
        dp[str.length()] = 1;
        for (int i = str.length() - 1; i >= 0; i--) {
            for (int j = i + 1; j <= Math.min(str.length(), i + 2); j++) {
                if (canTranslate(str, i, j)) {
                    dp[i] += dp[j];
                }
            }
        }
        return dp[0];
    }

    public boolean canTranslate(String num, int i, int j) {
        if (i >= j) return false;
        if (j - i > 1 && num.charAt(i) == '0') return false;
        int tmp = Integer.parseInt(num.substring(i, j));
        return tmp >= 0 && tmp <= 25;
    }

    public static void main(String[] args) {
        System.out.println(new TranslateNum_46().translateNum(12258));
    }
}
